app/policies/company_policy.rb

```diff
@@ -20,16 +18,17 @@ def create?
   end
 
   def update?
-    is_admin? || @user
+   is_admin? || is_in_company?
   end
 
 
-  def edit?
-    update?
-  end
 
   private
   def is_admin?
     @user.admin? if @user
   end
+
+  def is_in_company?
+    @user && @user.is_in_company_numbered?(@record.company_number)
```

Comment:
This might be more succinct as:
```
@user.&.is_in_company_numbered?(@record.company_number)
```
This uses the "safe navigation method": http://stackoverflow.com/questions/36812647/what-does-ampersand-dot-mean-in-ruby

And, with devise we are sure that @user will be at least defined.